storage/tsdb/tsdb.go

```diff
@@ -27,6 +28,63 @@ import (
 	tsdbLabels "github.com/prometheus/tsdb/labels"
 )
 
+// ErrNotReady is returned if the underlying storage is not ready yet.
+var ErrNotReady = errors.New("TSDB not ready")
+
+// ReadyStorage implements the Storage interface while allowing to set the actual
+// storage at a later point in time.
+type ReadyStorage struct {
+	mtx sync.RWMutex
+	a   *adapter
+}
+
+// Set the storage.
+func (s *ReadyStorage) Set(db *tsdb.DB) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.a = &adapter{db: db}
+}
+
+// Get the storage.
+func (s *ReadyStorage) Get() *tsdb.DB {
+	if x := s.get(); x != nil {
+		return x.db
+	}
+	return nil
+}
+
+func (s *ReadyStorage) get() *adapter {
+	s.mtx.RLock()
+	x := s.a
```

Comment:
Interesting. What's the reason to have this additional variable assignment? I'd have naively written:

```go
s.mtx.RLock()
defer s.mtx.RUnlock()
return s.a
```

Reply:
Yea, that's probably fine. I've just had a few issues in the last couple of months where `defer` had a very significant performance penalty. Just got in a habit of skipping on it if there was no legitimate concern of intermediate panics.

Reply:
Thanks for the info! Was just curious.